src/Database/Franz/Server.hs

```diff
@@ -0,0 +1,181 @@
+{-# LANGUAGE DeriveGeneric, LambdaCase, OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE DeriveFunctor #-}
+{-# LANGUAGE BangPatterns #-}
+module Database.Franz.Server
+  ( Settings(..)
+  , startServer
+  , defaultPort
+  ) where
+
+import Control.Concurrent
+import Control.Exception
+import Control.Monad
+import Control.Retry
+import Control.Concurrent.STM
+import Database.Franz.Internal
+import Database.Franz.Protocol
+import Database.Franz.Reader
+import Data.ConcurrentResourceMap
+import Data.Serialize
+import qualified Data.IntMap.Strict as IM
+import Data.IORef
+import qualified Data.ByteString.Char8 as B
+import qualified Data.HashMap.Strict as HM
+import Data.Tuple (swap)
+import qualified Network.Socket.SendFile.Handle as SF
+import qualified Network.Socket.ByteString as SB
+import qualified Network.Socket as S
+import System.Directory
+import System.FilePath
+import System.IO
+import System.Process (ProcessHandle, spawnProcess, cleanupProcess,
+  waitForProcess, getProcessExitCode, getPid)
+
+respond :: FranzReader
+  -> IORef (IM.IntMap ThreadId)
+  -> FilePath
+  -> IORef B.ByteString
+  -> MVar S.Socket -> IO ()
+respond env refThreads path buf vConn = do
+  recvConn <- readMVar vConn
+  runGetRecv buf recvConn get >>= \case
+    Right (RawRequest reqId req) -> do
+      let pop result = do
+            case result of
+              Left ex | Just e <- fromException ex -> sendHeader $ ResponseError reqId e
+              _ -> pure ()
+            void $ popThread reqId
```

Comment:
The `sendHeader` above could easily fail, right? It's network.

I'm not sure of consequence of it failing yet, but we can at least ensure we don't leak thread ID: use `finally`.

Reply:
It's not a huge issue as network failure implies cleaning up everything, but sure

Reply:
It's not a huge deal because a network failure implies terminating the request loop (and killing all threads), but sure